go-controller/pkg/ovn/controller/podset/controller.go

```diff
@@ -0,0 +1,423 @@
+// PodSets that were added during that time.
+func (psc *PodSetController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer psc.queue.ShutDown()
+
+	klog.Infof("Starting controller %s", controllerName)
+	defer klog.Infof("Shutting down controller %s", controllerName)
+
+	psc.podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    psc.onPodAdd,
+		UpdateFunc: psc.onPodUpdate,
+		DeleteFunc: psc.onPodDelete,
+	})
+	psc.namespaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    psc.onNamespaceAdd,
+		UpdateFunc: psc.onNamespaceUpdate,
+		DeleteFunc: psc.onNamespaceDelete,
+	})
+
+	// Wait for the caches to be synced
+	klog.Info("Waiting for informer caches to sync")
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, psc.podsSynced, psc.namespacesSynced) {
+		return fmt.Errorf("error syncing cache")
+	}
+
+	// sync any PodSets that arrived while we were waiting for informer sync
+	go psc.syncPendingPodSets(stopCh)
+
+	// begin processing
+	klog.Info("PodSetController: Starting workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(psc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	return nil
+}
+
+// syncPendingPodSets syncs any PodSets that arrived while we were waiting for
+// our informers to sync
+func (psc *PodSetController) syncPendingPodSets(stopCh <-chan struct{}) {
+	// cute: even though we're deleting from podSetsToSync, nothing will
+	// add to it, since by the time this function is called, the informers are synced
+	// and any new PodSets will be synced directly.
+	psc.podSetsLock.RLock()
+	defer psc.podSetsLock.RUnlock()
+
+	first := true
+	for len(psc.podSetsToSync) > 0 {
+		klog.Infof("Syncing %d pending PodSets", len(psc.podSetsToSync))
+		if !first {
+			time.Sleep(10 * time.Second)
+		}
+		for psname := range psc.podSetsToSync {
+			if err := psc.syncPodSet(psc.podSets[psname]); err != nil {
+				klog.Errorf("Failed to sync PodSet %s (will retry): %v", psname, err)
+			} else {
+				delete(psc.podSetsToSync, psname)
+			}
+
+			// abort if asked to stop
+			select {
+			case <-stopCh:
+				return
+			default:
+			}
+		}
+	}
+
+	klog.Infof("Done syncing all pending PodSets")
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same namespace or pod
+// at the same time.
+func (psc *PodSetController) worker() {
+	for psc.processNextWorkItem() {
+	}
+}
+
+func (psc *PodSetController) processNextWorkItem() bool {
+	eKey, quit := psc.queue.Get()
+	if quit {
+		klog.V(5).Infof("PodSetController: stopping worker")
+		return false
+	}
+	defer psc.queue.Done(eKey)
+	key := eKey.(string)
+	klog.V(5).Infof("PodSetController: processing key %s", key)
+
+	// split & validate key
```

Comment:
this kind of namespace manipulation is a little nasty - yes i did this to differentiate between add/update/delete in the pkg/informer but I really didn't like it.

your controller should have one type of resource that you are attempting to reconcile (i.e pods) and may have one or more secondary resources (i.e namespaces) that when changed have an affect on one or more primary resources.

i'm not well versed on how podsets work, but I would surmise that you are really working on a Pod resource. When a namespace changes your `onNamespaceAdd/Update` function will want to list on the pods in that namespace and then call `enqueuePod` for each pod affected.

Reply:
A PodSet is two label selectors: namespaces, and names. We want to mirror that in an OVN AddressSet object. Thus, we need to react to:

1. PodSet creation / deletion (it is derived from NetworkPolicy)
2. Pod changes - creation, deletion, address changes, label changes
3. Namespace changes - creation, deletion, and label changes.

> your controller should have ... one or more secondary resources

That's exactly what this is - the primary resource is a NetworkPolicy, and the policy controller manipulates PodSets. In a sense, this is the thing that watches the secondary resources.

We could enqueue PodSet identifiers rather Pods or Namespaces. That would match the controller model a bit better. However, one of the goals is to synchronously add Pods to all their destination address sets on create, so we never have a pod without NetworkPolicy applied. That sot of requires us to commit on Pod updates.

If you have another idea, I'm all ears.

Reply:
> That's exactly what this is - the primary resource is a NetworkPolicy, and the policy controller manipulates PodSets. In a sense, this is the thing that watches the secondary resources.

Then I would have expected there to be a watch for NetworkPolicy resources in this controller. It would be better for this controller to work in isolation from the main policy controller IMHO. Your Pod and Namespace add/update/delete events would then trigger an update for whichever network policies are applied.

> We could enqueue PodSet identifiers rather Pods or Namespaces. That would match the controller model a bit better. However, one of the goals is to synchronously add Pods to all their destination address sets on create, so we never have a pod without NetworkPolicy applied. That sot of requires us to commit on Pod updates.

That's a tough ask considering we can only guarantee eventual consistency.

> If you have another idea, I'm all ears.

Hmmmm.... I would have our CNI server wait for a signal that the pod network is ready before returning...
Then I'd have a NetworkReadiness controller, which listens to K8s pod events to see that all the ready conditions (i.e annotations from other parts of the controller) are set, at which point, I'd signal the CNI plugin that we're good to return.
Sure, network attachment might take longer, but you'd be guaranteed you wouldn't create a pod without all policies being applied.

Reply:
> That's a tough ask considering we can only guarantee eventual consistency.

We can guarantee whatever we want - we control pod ipam :-).

Right now, this controller reacts to both informer events, as well as being called directly from addLogicalPort. That means that policy on pod creation is now synchronous, removing a decent source of bugs.

I would be open to making a larger Pod controller, but step by tiny step.

Reply:
> We can guarantee whatever we want - we control pod ipam
:joy: 

> I would be open to making a larger Pod controller, but step by tiny step.

*cough* #2096 *cough*

> Right now, this controller reacts to both informer events, as well as being called directly from addLogicalPort

Yeah... that entangles this controller with the pod one I'm working on so I'm :-1: on adding calls through `addLogicalPort`. It's already very tangled with exGW and I'm trying to get that factored out in to it's own controller too.